lib/Semantics/resolve-names.cpp

```diff
@@ -3821,14 +3821,36 @@ void DeclarationVisitor::Post(const parser::ComponentDecl &x) {
       !attrs.HasAny({Attr::PUBLIC, Attr::PRIVATE})) {
     attrs.set(Attr::PRIVATE);
   }
-  if (!attrs.HasAny({Attr::POINTER, Attr::ALLOCATABLE})) {
-    if (const auto *declType{GetDeclTypeSpec()}) {
-      if (const auto *derived{declType->AsDerived()}) {
+  if (const auto *declType{GetDeclTypeSpec()}) {
+    if (const auto *derived{declType->AsDerived()}) {
+      if (!attrs.HasAny({Attr::POINTER, Attr::ALLOCATABLE})) {
         if (derivedTypeInfo_.type == &derived->typeSymbol()) { // C744
           Say("Recursive use of the derived type requires "
               "POINTER or ALLOCATABLE"_err_en_US);
         }
       }
+      if (!coarraySpec().empty()) { // C747
+        if (IsTeamType(derived)) {
+          Say("A coarray component may not be of type TEAM_TYPE from "
+              "ISO_FORTRAN_ENV"_err_en_US);
+        } else {
+          if (IsIsoCType(derived)) {
+            Say("A coarray component may not be of C_PTR or C_FUNPTR from "
+                "ISO_C_BINDING when an allocatable object is a "
+                "coarray"_err_en_US);
+          }
+        }
+      }
+      if (FindCoarrayUltimateComponent(*derived)) { // C748
+        if (attrs.HasAny({Attr::POINTER, Attr::ALLOCATABLE})) {
+          Say("A component whose type has a coarray ultimate component "
+              "cannot be a POINTER or ALLOCATABLE"_err_en_US);
```

Comment:
We have mostly been using "may not" rather than "cannot".

Reply:
I also use "must not", but I shouldn't -- "may" is the correct modal verb here, since we're talking about what's allowed, not what might be possible.

You might as well capture the result of `FindCoarrayUltimateComponent` and use its name in the message and link to its declaration.

Reply:
Thanks.  I'll fix it.

Reply:
I notice that there are many uses of the word "cannot" in resolve-names.cpp.  @tskeith, should I fix them all in this pull request?

Reply:
I think it makes more sense to change them in a separate pull request, but if you want to do it in this one that's okay too.

Reply:
I think I'll wait.